flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java

```diff
@@ -269,6 +288,29 @@ void onSuccessfulCompletionOfCheckpoint(long checkpointId) throws IOException {
 				it.remove();
 			}
 		}
+
+		cleanupOutdatedResumables(checkpointId);
+	}
+
+	private void cleanupOutdatedResumables(long checkpointId) throws IOException {
+		Iterator<Map.Entry<Long, RecoverableWriter.ResumeRecoverable>> it =
```

Comment:
If this is the place that you were referring earlier for optimization, that I'd also suggest to consider [TreeMap#headMap](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#headMap-K-boolean-) instead of HashMap, but I suspect that in reality the number of elements would be relatively low, wouldn't it?

Reply:
I would assume so, but it cannot be guaranteed. So I would say that your recommendation makes sense.